python/ray/data/_internal/split.py

```diff
@@ -0,0 +1,237 @@
+        split_single_block_results = ray.get(split_single_block_futures)
+        for block_id, block_split_result in split_single_block_results:
+            all_blocks_split_results[block_id] = block_split_result
+    return all_blocks_split_results
+
+
+def _merge_all_blocks_split_results(
+    all_blocks_split_results: List[List[Tuple[ObjectRef[Block], BlockMetadata, int]]],
+) -> Tuple[List[List[ObjectRef[Block]]], List[List[BlockMetadata]]]:
+    """Merge per block's split result into final split result."""
+    result_blocks = []
+    result_metas = []
+    current_blocks = []
+    current_meta = []
+    # the merge algorithm is simple: we only need to create a new
+    # split if the current block has more than 1 split.
+    for single_block_split_result in all_blocks_split_results:
+        for i, (block, meta, _) in enumerate(single_block_split_result):
+            if i != 0:
+                result_blocks.append(current_blocks)
+                result_metas.append(current_meta)
+                current_blocks = []
+                current_meta = []
+            current_blocks.append(block)
+            current_meta.append(meta)
+    result_blocks.append(current_blocks)
+    result_metas.append(current_meta)
+
+    return result_blocks, result_metas
+
+
+def _split_at_indices(
+    blocks_with_metadata: Iterable[Tuple[ObjectRef[Block], BlockMetadata]],
+    indices: List[int],
+) -> Tuple[List[List[ObjectRef[Block]]], List[List[BlockMetadata]]]:
+    """Split blocks at the provided indices.
+
+    Args:
+        blocks_with_metadata: Block futures to split, including the associated metadata.
+        indices: The (global) indices at which to split the blocks.
+    Returns:
+        The block split futures and their metadata. If an index split is empty, the
+        corresponding block split future will be None.
+    """
+
+    # We implement the split in 3 phases.
+    # phase 1: calculate the per block split indices.
+    blocks_with_metadata = list(blocks_with_metadata)
+    block_sizes: List[int] = _calculate_blocks_size(blocks_with_metadata)
+    valid_indices = _generate_valid_indices(block_sizes, indices)
+    per_block_split_indices: List[List[int]] = _generate_per_block_split_indices(
+        block_sizes, valid_indices
+    )
+
+    # phase 2: split each block based on the indices from previous step.
+    all_blocks_split_results: List[
+        List[Tuple[ObjectRef[Block], BlockMetadata, int]]
+    ] = _split_all_blocks(blocks_with_metadata, block_sizes, per_block_split_indices)
+
+    # phase 3: generate the final split.
+    return _merge_all_blocks_split_results(all_blocks_split_results)
+
+
+def _get_num_rows(block: Block) -> int:
+    """Get the number of rows contained in the provided block."""
+    return BlockAccessor.for_block(block).num_rows()
+
+
+def _split_at_index(
+    blocks_with_metadata: Iterable[Tuple[ObjectRef[Block], BlockMetadata]],
+    index: int,
+    return_right_half: bool,
```

Comment:
return_right_half = False is only used for dataset.limit which I think not worth optimizing for now. And to be honest the current implementation is not too bad: we just create an extra copy of the block being splitted, which is minimal.

Reply:
We are actually working on eager memory releasing of the input block that's been split: https://github.com/ray-project/ray/pull/26583